### PR TITLE
control_mpc: add MPC control law

### PR DESCRIPTION
**Pull request by dalegaard**
This adds MPC control law to main branch.

Has been extensively tested on `feature/mpc_experimental` branch but would like to get to main so we can get wider testing. Would appreciate final code review.

Let me know if we need to add any extra docs.

Best regards,
Lasse

### CODE REVIEW

**Comment by YanceyA** [comment]
@dalegaard I am working on a doc update from my own branch. It should be ready for PR this weekend. Let me know if you specifically have any areas which need to be added, changed or improved.

https://github.com/YanceyA/Danger_Klipper_Docs/blob/doc_mpc_update_july/docs/MPC.md

Rough TODO:
-Adjust docs to have "experimental section" with bed function details. Remove references to BED in main section of the doc for clarity. 
-Add PTC temp-power reference table 
-Add any relevent discord notes, specifically around ambient temp parameter.
-Remove ambient temp parameter from primary config section to optional parameters to mitigate people placing too much emphasis on it
-Add temp wait macro
-Clean up feed forward macro section 
-Add notes that MPC controls block temp and not sensor temp to mitigate user confusion.

**Comment by rogerlz** [comment]
thanks @YanceyA. I added a bit of doc to the config reference. Let me know if you want me to correct or add anything else there

File: docs/Config_Reference.md [hunk 1]
```diff
@@ -1037,11 +1040,27 @@ pid_Kd:
...
+#
+#   If control: watermark
 #max_delta: 2.0
 #   On 'watermark' controlled heaters this is the number of degrees in
 #   Celsius above the target temperature before disabling the heater
 #   as well as the number of degrees below the target before
 #   re-enabling the heater. The default is 2 degrees Celsius.
+#
+#   If control: mpc
+#   See MPC.md for details about these parameters.
```
Comment (YanceyA) [inline docs/Config_Reference.md:1052]: Hi, i suggest to use only the following config options here as the minimum set to get people operational. 
#heater_power:
#cooling_fan:
#filament_diameter: 1.75
#filament_density: 1.20
#filament_heat_capacity: 1.8 

The following five parameters are advanced settings that most users won't need to configure beyond the build in defaults. The add a bit of config bloat if included. I think they should be removed here.  
- ambient_temp_sensor  
- target_reach_time  
- smoothing  
- min_ambient_change  
- steady_state_rate

**Review by YanceyA — COMMENTED** [review]
Roger, a few small comments on the config. Otherwise it all looks great.

**Comment by rogerlz** [comment]
> Roger, a few small comments on the config. Otherwise it all looks great.

thanks for the suggestions

**Review by rogerlz — APPROVED** [review]

**Review by rogerlz — APPROVED** [review]

**Review by rogerlz — APPROVED** [review]

**Pull request merged**